### PR TITLE
UnorderedKey fix

### PR DESCRIPTION
**Pull request by evgeniy-r**
PR for https://github.com/dotenv-linter/dotenv-linter/issues/252.

#### ✔ Checklist:

- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);
- [x] Tests for the changes have been added (for bug fixes / features);
- [x] Docs have been added / updated (for bug fixes / features).

### CODE REVIEW

**Comment by baile320** [comment]
Quick question: should this support key sorting within groupings?

For example:
```
# Grouping 1
A=B
C=B
B=D

# Grouping 2
THIS=THAT
Z=X
Q=R
```

after running the fixer could look like:

```
# Grouping 1
A=B
B=D
C=B

# Grouping 2
Q=R
THIS=THAT
Z=X
```

When I run the current fixer, I get:

```
# Grouping 1
A=B
B=D
C=B

Q=R
# Grouping 2
THIS=THAT
Z=X
```
which is not behavior I would expect if I were a user because I think users frequently group common functionality together in env files (all database connections go near each other, all message broker connections near each other, etc). If it is decided to keep the functionality as it currently is, it might be helpful to add explanations in the doc somewhere. Currently, the docs say "You can use blank lines to split lines into groups" (https://github.com/dotenv-linter/dotenv-linter#unordered-key) so that might lead to confusion.

Open to hearing other people's thoughts though.

**Comment by evgeniy-r** [comment]
This behavior is due to the fact that we can handle comments in different ways.

I assumed that the comment is most likely related to the next line with a key (`# Grouping 2` is related to `THIS=THAT`).
So it is handled as a whole.

It can be disscussed and changed if necessary.

**Comment by evgeniy-r** [comment]
And yes, we sort the keys separately in each group.

**Comment by mgrachev** [comment]
@evgeniy-r Thank you for your contribution! 🔥

I need more time to review this PR 🙂

File: src/fixes/unordered_key.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    ) -> Option<usize> {
+        // At first we find all groups
+
+        let mut sorted_part_started = false;
+        let mut start = None;
+        let mut end = None;
+
+        let mut sorted_groups = Vec::with_capacity(lines.len());
+
+        for (i, line) in lines.iter().enumerate() {
```
Comment (DDtKey) [inline src/fixes/unordered_key.rs:48]: I think that it is worth changing the approach here, the abundance of `Option` variables complicates the logic and requires unwanted `expect()`:thinking:

Maybe something like this? (quickly written code, may contain an error, just an attempt to convey my idea)
```
        let mut start_part_index = 0;

        for i in 0..lines.len()  {
            match lines.get(i) {
                Some(el) if el.is_empty() => {
                    if start_part_index < i {
                        let part = &mut lines[start_part_index..i];
                        sort_part(part);
                    }
                    start_part_index = i + 1;
                }
                _ => {}
            }
        }
```

This should be consistent with my other comments :slightly_smiling_face:

**Review by DDtKey — COMMENTED** [review]
You did a great job and thank you for that! :+1: 

I left some related comments, please take a look.

This is by no means a criticism, just considering options :sweat_smile:

**Comment by DDtKey** [comment]
Among other things, I noticed that we work exclusively with `lines` (**without checking warnings!**) and we always sort them.
But we have control comments(#237) and not the entire file should be affected by this `Fixer`, but only really existing `Warning`s.

I understand that the order can break, but this also does not seem to be a suitable solution.
And it's really trouble... 
I also found this in the recently added `EndingBlankLineFixer` (#263)

Perhaps this is beyond the scope of this task at the moment, but we cannot forget about it.

What do you think @dotenv-linter/core @evgeniy-r ?

**Comment by mgrachev** [comment]
> Among other things, I noticed that fixer works exclusively with `lines` (**without checking warnings!**) and always sort full file.
> But we have control comments(#237) and not the entire file should be affected by this `Fixer`, but only really existing `Warning`s.

Good catch! You're absolutely right. All checks and fixes should be disabled using controll comments or the `--skip` argument.

> I understand that the warning `line_number` can break in fix process, but this also does not seem to be a suitable solution.
> And it's really trouble...
> I also found this in the recently added `EndingBlankLineFixer` (#263)

Can you give more information about the problem with `EndingBlankLineFixer` and controll comments?
I've just checked it and it's worked with the controll comment `# dotenv-linter:off EndingBlankLine`.
Or you can open an issue and we will discuss about it.

> One option is of course to check the control comments in the fixer. But this will create duplicate logic.
> At the same time, changing `line_number` in `Warning`s is also not correct, because when displaying results, we must point to the line of error in the original file (but when there are implementations of all fixers, we may not print this at all).
> 
> Perhaps this is beyond the scope of this task at the moment, but we cannot forget about it.

No, I don't think so. We should find a solution before merging this PR 🤔

**Comment by DDtKey** [comment]
> Can you give more information about the problem with `EndingBlankLineFixer` and controll comments?
> I've just checked it and it's worked with the controll comment `# dotenv-linter:off EndingBlankLine`.
> Or you can open an issue and we will discuss about it.

Sorry, I was wrong (typo), I meant `ExtraBlankLineFixer` (#260) and [this review comment](https://github.com/dotenv-linter/dotenv-linter/pull/260#discussion_r471299204)
A little later, I can double-check and create an issue.

> No, I don't think so. We should find a solution before merging this PR thinking

I fully agree.

**Comment by mgrachev** [comment]
> Sorry, I was wrong (typo after GH suggest), I meant `ExtraBlankLineFixer` (#260) and [this review comment](https://github.com/dotenv-linter/dotenv-linter/pull/260#discussion_r471299204)
> A little later, I can double-check and create an issue.

Let me know. Thank you 🙏

**Comment by evgeniy-r** [comment]
I was away, will answer all comments soon.

File: src/fixes/unordered_key.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        for warning in warnings {
+            warning.mark_as_fixed();
+        }
+
+        Some(count)
+    }
+}
+
+impl UnorderedKeyFixer<'_> {
+    fn sort_part(lines: &mut [LineEntry], start: usize, end: usize) {
```
[the remark this answers is not in the document]
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs:84]: To make sure we are talking about the same thing.

We don't make changes within the slices.
Each slice is a significant line (with key) + previous comments (if present). Most of slices consist of one line.

[the remark this answers is not in the document]
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

File: src/fixes/unordered_key.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        warnings: Vec<&mut Warning>,
+        lines: &mut Vec<LineEntry>,
+    ) -> Option<usize> {
+        // At first we find all groups
+
+        let mut sorted_part_started = false;
+        let mut start = None;
+        let mut end = None;
+
+        let mut sorted_groups = Vec::with_capacity(lines.len());
```
[the remark this answers is not in the document]
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: I have removed `sorted_groups`.

File: src/fixes/unordered_key.rs (hunk 1 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: I didn't quite get the idea (

We don't need `match` in this code snippet, we can use `[]`.

And it doesn't cover all cases (e.g. with comments).

There are 3 variables (`sorted_part_started`, `start`, `end`) and it seems that we need all of them.
A comment may or may not be the start of a sorted group:
```

# This is not
# a sorted group.
# No keys, just comments

```

Also `i` is not always the end of a sorted group:
```

A=B
C=D
# The end of a sorted group is
# the line `C=D`,
# not this one

```

**Comment by evgeniy-r** [comment]
Before discussing control comments, I have a question.
Is this the expected behavior?

File:
```
X=Y
# dotenv-linter:off UnorderedKey
A1=B
A2=B

# dotenv-linter:on UnorderedKey
A=B
```

Linter's output:
```
.env:7 UnorderedKey: The A key should go before the X key

Found 1 problem
```

**Comment by DDtKey** [comment]
> Before discussing control comments, I have a question.
> Is this the expected behavior?

I think there is some logic in this.
Actually, the group is separated in the comment block and it (block) is not surrounded by blank lines
But it's may be not clear.

But from my point of view, this is the expected result, and in this case, the following option looks correct for separation:
```
X=Y
# dotenv-linter:off UnorderedKey
A1=B
A2=B
# dotenv-linter:on UnorderedKey

A=B
```

That is, I would not consider comments attached to specific lines if they are separated from them(key\value line) by a line break :thinking: 

Maybe @mgrachev has a different vision on this?

File: src/fixes/unordered_key.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                }
+                if !line.is_comment() {
+                    end = Some(i + 1);
+                    sorted_part_started = true;
+                }
+            }
+        }
+
+        // The last group
+        if sorted_part_started {
```
Comment (DDtKey) [inline src/fixes/unordered_key.rs:68]: What about just add in the body of the `for` cycle additional condition for last group instead of checking after the loop?
For example, in `for`:
```
if line.is_empty() || (i == lines.len() - 1)
```

Comment (evgeniy-r) [reply src/fixes/unordered_key.rs:68]: In this case, we need to add more conditions, since the `end` variable has not been assigned a value `lines.len() - 1`.

Comment (DDtKey) [reply src/fixes/unordered_key.rs:68]: Yeah, you're right, sorry :slightly_smiling_face:

File: src/fixes/unordered_key.rs (hunk 1 above)
Comment (DDtKey) [reply src/fixes/unordered_key.rs]: In fact, 3 variables seem redundant to me. I apologize that the code snippet was not clear.

Suppose in your version, the _start index_ is always shifted by 1 after finding an empty string (in next iteration). We may not use `Option` here.

In the case of `sorted_part_started`, I see a clear correlation with the `end`.

How do you like this option? He must pass all tests.

```
        let mut start = 0;
        let mut end = None;

        for i in 0..lines.len() {
            let line = &lines[i];
            let is_last_line = i == lines.len() - 1;

            if line.is_empty() || is_last_line {
                if let Some(end_index) = end {
                    Self::sort_part(
                        &mut lines[start..end_index],
                    );
                    end = None;
                }
                start = i + 1;
            } else if !line.is_comment() {
                end = Some(i + 1);
            }
        }
```

Perhaps this can be done even more expressively. But this is still the main idea.

Also, this option saves us from the need to check the last group outside the loop.

File: src/fixes/unordered_key.rs (hunk 4 above)
Comment (DDtKey) [reply src/fixes/unordered_key.rs:68]: Please see [this comment](https://github.com/dotenv-linter/dotenv-linter/pull/261#discussion_r481022360), it solves this issue

Comment (DDtKey) [reply src/fixes/unordered_key.rs:68]: Please see [this comment](https://github.com/dotenv-linter/dotenv-linter/pull/261#discussion_r481022360), it solves this issue

File: src/fixes/unordered_key.rs (hunk 1 above)
Comment (DDtKey) [reply src/fixes/unordered_key.rs]: I also noticed that we seem to be missing a test for something where the last line is not empty, but here we will get two `Warning`s (`UnorderedKey` and `EndingBlankLine`)

For example:
```dotenv
       # I comment\n\
        I=123\n\
        \n\
        B1=1\n\
        A1=1",
    );
```
Maybe add an integration test?


Therefore, to check the last line fully correct, we can slightly change the end condition:
```
        let mut start = 0;
        let mut end = None;

        for i in 0..lines.len() {
            let line = &lines[i];
            let is_last_line = i == lines.len() - 1;

            if !line.is_empty_or_comment(){
                end = Some(i + 1);
            }

            if line.is_empty() || is_last_line {
                if let Some(end_index) = end {
                    Self::sort_part(
                        &mut lines[start..end_index],
                    );
                    end = None;
                }
                start = i + 1;
            }
        }
```

File: src/fixes/unordered_key.rs (hunk 4 above)
Comment (DDtKey) [reply src/fixes/unordered_key.rs:68]: > In this case, we need to add more conditions, since the `end` variable has not been assigned a value `lines.len() - 1`.

Current tests cannot cover this condition anyway, they always have `\n` at the end. Please see also [comment](https://github.com/dotenv-linter/dotenv-linter/pull/261#discussion_r481040776)

**Comment by DDtKey** [comment]
> > Sorry, I was wrong (typo after GH suggest), I meant `ExtraBlankLineFixer` (#260) and [this review comment](https://github.com/dotenv-linter/dotenv-linter/pull/260#discussion_r471299204)
> > A little later, I can double-check and create an issue.
> 
> Let me know. Thank you

I created an issue #278 with examples :slightly_smiling_face:

**Comment by mgrachev** [comment]
> At the same time, changing `line_number` in `Warning`s is also not correct, because when displaying results, we must point to the line of error in the original file (but when there are implementations of all fixers, we may not print this at all).

I think we should remove the logic for displaying `Fixed/Unfixed Warnings` after we merge this PR.

**Comment by evgeniy-r** [comment]
> 
> I think there is some logic in this.
> Actually, the group is separated in the comment block and it (block) is not surrounded by blank lines. But it's may be not clear.
> 

So, the answer is "yes" (that is the desired behavior)? :slightly_smiling_face:

**Comment by mgrachev** [comment]
> So, the answer is "yes" (that is the desired behavior)? 🙂

Yes, the `UnorderedKey` check works as expected.

File: src/fixes/unordered_key.rs (hunk 1 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: This solution is simpler, thank you.
I have used it (I have rename `start` to `start_index`) and have added the test (unit kind, I think this is OK).

File: src/fixes/unordered_key.rs (hunk 4 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

**Comment by evgeniy-r** [comment]
> Yes, the UnorderedKey check works as expected.

I plan to submit a PR tomorrow (control comments).

**Comment by evgeniy-r** [comment]
I found an interesting kind of issues:

Before fixing the order:
```
BB=1
# dotenv-linter:off LowercaseKey
Aa=B
X=X
# dotenv-linter:on LowercaseKey

```

After fixing:
```
Aa=B
BB=1
# dotenv-linter:off LowercaseKey
X=X
# dotenv-linter:on LowercaseKey

```
And we will get the incorrect file.
If we fix the `Aa` key somehow, it will not be the best solution either, because a user doesn't want fix the lower case.

Therefore, I believe that we should split the groups by control comments, as well as by blank lines.

I made appropriate changes.

**Comment by evgeniy-r** [comment]
I moved changes in checker to the separate PR: https://github.com/dotenv-linter/dotenv-linter/pull/281

File: src/fixes/unordered_key.rs [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+                is_off = comment.is_disabled();
+            }
+
+            if !is_disabled {
+                if !line.is_empty_or_comment() {
+                    end.replace(i + 1);
+                }
+
+                let is_last = i == lines.len() - 1;
+                if line.is_empty() || is_last || is_control_comment {
```
Comment (mgrachev) [inline src/fixes/unordered_key.rs:66]:
```suggestion
                if line.is_empty() || line.is_last_line() || is_control_comment {
```
Why don't you use the `is_last_line` method?

File: src/fixes/unordered_key.rs [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+            if !line.is_comment() {
+                slices.push(&part[slice_start..=i]);
+                slice_start = i + 1;
+            }
+        }
+
+        slices.sort_by_cached_key(|slice| {
+            // I think, that we should modify get_key() so it will return Option<&str> instead of
+            // Option<String>.
+            slice.last().expect("significant line").get_key()
```
Comment (mgrachev) [inline src/fixes/unordered_key.rs:105]:
```suggestion
            slice.last()?.get_key()
```

File: src/fixes/unordered_key.rs [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+mod tests {
+    use super::*;
+    use crate::common::tests::line_entry;
+
+    fn get_lines(strs: Vec<&str>) -> Vec<LineEntry> {
+        let total = strs.len();
+
+        strs.iter()
+            .enumerate()
+            .map(|(i, &str)| line_entry(i + 1, total, str))
```
Comment (mgrachev) [inline src/fixes/unordered_key.rs:127]:
```suggestion
            .map(|(i, &line)| line_entry(i + 1, total, line))
```
Please don't use `str` as a variable name, because Rust has the same type's name - [str](https://doc.rust-lang.org/std/primitive.str.html).

File: src/fixes/unordered_key.rs [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::common::tests::line_entry;
+
+    fn get_lines(strs: Vec<&str>) -> Vec<LineEntry> {
+        let total = strs.len();
+
+        strs.iter()
```
Comment (mgrachev) [inline src/fixes/unordered_key.rs:125]:
```suggestion
    fn get_lines(lines: Vec<&str>) -> Vec<LineEntry> {
        let total = lines.len();

        lines.iter()
```

File: src/fixes/unordered_key.rs [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+            .enumerate()
+            .map(|(i, &str)| line_entry(i + 1, total, str))
+            .collect()
+    }
+
+    fn get_warnings(lines: &[LineEntry], wrns: Vec<(usize, &str)>) -> Vec<Warning> {
+        wrns.into_iter()
+            .map(|(i, str)| Warning::new(lines[i].clone(), "UnorderedKey", String::from(str)))
```
Comment (mgrachev) [inline src/fixes/unordered_key.rs:133]:
```suggestion
    fn get_warnings(lines: &[LineEntry], warnings: Vec<(usize, &str)>) -> Vec<Warning> {
        warnings.into_iter()
            .map(|(i, line)| Warning::new(lines[i].clone(), "UnorderedKey", String::from(line)))
```

File: src/fixes/unordered_key.rs [hunk 10]
```diff
@@ -0,0 +1,724 @@
...
+    fn run_fixer(warnings: &mut [Warning], lines: &mut Vec<LineEntry>) -> Option<usize> {
+        let mut fixer = UnorderedKeyFixer::default();
+        let warning_refs = warnings.iter_mut().collect();
+
+        fixer.fix_warnings(warning_refs, lines)
+    }
+
+    fn assert_lines(lines: &[LineEntry], strs: Vec<&str>) {
+        for (i, &str) in strs.iter().enumerate() {
+            assert_eq!(str, lines[i].raw_string.as_str());
```
Comment (mgrachev) [inline src/fixes/unordered_key.rs:146]:
```suggestion
    fn assert_lines(result: &[LineEntry], lines: Vec<&str>) {
        for (i, &line) in lines.iter().enumerate() {
            assert_eq!(line, result[i].raw_string.as_str());
```

File: tests/common/mod.rs [hunk 11]
```diff
@@ -128,6 +128,7 @@ impl TestDir {
     }
 
     /// Run the default CLI binary, with "-f", in this TestDir and check it fails.
+    #[allow(dead_code)]
```
Comment (mgrachev) [inline tests/common/mod.rs:131]: Please remove this method.

File: tests/fixes/mod.rs [hunk 12]
```diff
@@ -54,30 +54,96 @@ fn key_without_value() {
 }
 
 #[test]
-fn unfixed_warnings() {
+fn unordered_key() {
```
Comment (mgrachev) [inline tests/fixes/mod.rs:57]: Please move this test to the `unordered_key.rs` file.

File: tests/fixes/mod.rs [hunk 13]
```diff
@@ -54,30 +54,96 @@ fn key_without_value() {
...
+        A1=1\n\
+        B1=1\n\
+        \n\
+        # end comment\n",
     );
 
     testdir.close();
 }
 
+// We need to figure out how to test unfixed warnings now when all the fixers have been implemented
```
Comment (mgrachev) [inline tests/fixes/mod.rs:119]: Please remove this test.

Comment (evgeniy-r) [reply tests/fixes/mod.rs]: Done

File: tests/fixes/mod.rs (hunk 12 above)
Comment (evgeniy-r) [reply tests/fixes/mod.rs]: Done

File: tests/common/mod.rs (hunk 11 above)
Comment (evgeniy-r) [reply tests/common/mod.rs]: Done

File: src/fixes/unordered_key.rs (hunk 10 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

File: src/fixes/unordered_key.rs (hunk 9 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

File: src/fixes/unordered_key.rs (hunk 8 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

File: src/fixes/unordered_key.rs (hunk 7 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

File: src/fixes/unordered_key.rs (hunk 6 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

File: src/fixes/unordered_key.rs (hunk 5 above)
Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Now we don't delete lines (just mark them), so this is OK.

Comment (evgeniy-r) [reply src/fixes/unordered_key.rs]: Done

**Review by mgrachev — APPROVED** [review]
